src/shared/settingsConfiguration.ts

```diff
@@ -5,231 +5,590 @@
-        return val ?? defaultValue
+            if (!section || e.affectsConfiguration(section, scope)) {
+                emitter.fire({ affectsConfiguration })
+            }
+        })
+
+        return [emitter, listener]
     }
 
     /**
-     * Gets the value of a VSCode setting.
+     * Defines a new set of configurables from the specified section.
      *
-     * @param key Setting name.
-     * @param type Setting type.
-     * @param opt.silent When to silently return undefined (default: "notfound"). notfound/undefined="nonexistent setting is not an error", yes="always silent", no="always error"
-     * @param opt.logging Enable logging (default: true).
+     * Only use this for creating 'virtual' settings that don't exist within `package.json`.
+     * Prefer {@link fromPackage} for most use-cases.
      */
-    public getSetting<T>(
-        key: string,
-        type: JSPrimitiveTypeName = 'string',
-        opt: {
-            silent?: 'no' | 'yes' | 'notfound'
-            logging?: boolean
-        } = {
-            silent: 'notfound',
-            logging: true,
+    public static define<T extends TypeDescriptor>(section: string, descriptor?: T) {
+        return createSettingsClass(section, descriptor)
+    }
+}
+
+// Keeping a separate function without a return type allows us to infer protected methods
+// TODO(sijaden): we can make this better in TS 4.7
+function createSettingsClass<T extends TypeDescriptor>(section: string, descriptor?: T) {
+    type Inner = FromDescriptor<T>
+
+    // Class names are not always stable, especially when bundling
+    const isMangled = SettingsConfiguration.name !== 'SettingsConfiguration'
+    function makeLogger(name = 'Settings') {
+        const prefix = `${isMangled ? 'Settings' : name} (${section})`
+        return (message: string) => getLogger().verbose(`${prefix}: ${message}`)
+    }
+
+    return class AnonymousSettings implements TypedSettings<Inner> {
+        private readonly config = this.settings.getSection(section)
+        private readonly onDidChangeEmitter = new vscode.EventEmitter<{ readonly key: keyof T }>()
+        private readonly disposables: vscode.Disposable[] = []
+        protected readonly log = makeLogger(Object.getPrototypeOf(this)?.constructor?.name)
+
+        public constructor(
+            private readonly settings: ClassToInterfaceType<SettingsConfiguration> = new SettingsConfiguration()
+        ) {
+            this.disposables.push(this.onDidChangeEmitter, this.registerChangedEvent(keys(descriptor ?? {})))
+        }
+
+        private registerChangedEvent(props: [keyof Inner & string]): vscode.Disposable {
+            const [sectionEmitter, listener] = this.settings.createScopedEmitter(section)
+            const sectionListener = sectionEmitter.event(event => {
+                for (const key of props.filter(p => event.affectsConfiguration(p))) {
+                    this.onDidChangeEmitter.fire({ key })
```

Comment:
an small example of a concrete case would make this easy to understand. e.g. 

    // for aws.foo.bar, section="..." and key="..."